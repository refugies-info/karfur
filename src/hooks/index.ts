export { default as useStateOnce } from "./useStateOnce";
