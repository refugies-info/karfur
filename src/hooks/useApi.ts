import axios from "react-native-axios";
import AsyncStorage from "@react-native-async-storage/async-storage";
import {
  UseQueryOptions,
  UseQueryResult,
  QueryKey,
  useQuery,
  useMutation,
  UseMutationOptions,
  UseMutationResult,
  MutationKey,
} from "react-query";
import "react-native-get-random-values"; // Needed before uuid import according to their docs
import { v4 as uuidv4 } from "uuid";

import Config from "../libs/getEnvironment";
import { logger } from "../logger";

type Method = "GET" | "POST" | "PUT" | "DELETE";

const dbURL = Config.dbUrl;
const siteSecret = Config.siteSecret;
const apiCaller = axios.create({
  baseURL: dbURL,
  headers: {
    "Content-Type": "application/json",
    "site-secret": siteSecret,
  },
});

const UID_STORAGE_KEY = "uid";

export const getUid = async (): Promise<string> => {
  const uid = await AsyncStorage.getItem("uid");
  if (uid) {
    return uid;
  }
  const newUid = uuidv4();
  await AsyncStorage.setItem(UID_STORAGE_KEY, newUid);
  return newUid;
};

export const makeApiRequest = async <T extends unknown>(
  url: string,
  payload: any,
  method: Method = "GET"
): Promise<T> => {
  try {
    let headers = {};

    const uid = await getUid();
    if (uid) {
      headers = {
        "x-app-uid": uid,
      };
    }

    let resp = null;

    if (method === "GET") {
      resp = await apiCaller.get(url, { headers, params: payload });
    } else if (method === "POST") {
      resp = await apiCaller.post(url, payload, { headers });
    }

    return resp?.data || <T>null;
  } catch (err) {
    logger.error(`makeApiRequest error: ${err}`);
    throw err;
  }
};

export const useApi = <Type, Error>(
  url: string,
  method: Method,
  key: string | any[],
  payload?: any,
  options?: UseQueryOptions<Type, Error, Type, QueryKey>
): UseQueryResult<Type, Error> => {
  return useQuery<Type, Error, Type>(
    key,
    () => makeApiRequest(url, payload, method),
    options as Omit<
      UseQueryOptions<Type, Error, Type, QueryKey>,
      "queryKey" | "queryFn"
    >
  );
};

export const useApiMutation = <Type, Error>(
  url: string,
  method: Method,
  key: string | any[],
  options?: UseMutationOptions<Type, Error, Type, MutationKey>
): UseMutationResult<Type, Error, Type> => {
  return useMutation<Type, Error, Type>(
    key,
    (payload: any) => makeApiRequest(url, payload, method),
    options as Omit<
      UseMutationOptions<Type, Error, Type>,
      "mutationKey" | "mutationFn"
    >
  );
};
