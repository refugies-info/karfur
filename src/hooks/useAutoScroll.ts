import { MutableRefObject, useEffect } from "react";
import { ScrollView } from "react-native";
import { useSelector } from "react-redux";
import { currentItem } from "../services/redux/VoiceOver/voiceOver.selectors";

export const useAutoScroll = (
  scrollviewRef: MutableRefObject<ScrollView | null>,
  offset: number
) => {
  const currentReadingItem = useSelector(currentItem);
  useEffect(() => {
    if (scrollviewRef && currentReadingItem) {
      scrollviewRef.current?.scrollTo({
        y: currentReadingItem.posY - offset,
        animated: true
      })
    }
  }, [currentReadingItem]);
}