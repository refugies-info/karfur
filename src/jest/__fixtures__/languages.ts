export const mockedLanguageData = [
  {
    _id: "5ce57c969aadae8734c7aedd",
    avancementTrad: undefined,
    i18nCode: "fr",
    langueFr: "Fran√ßais",
  },
  {
    _id: "5ce57c969aadae8734c7aee0",
    avancementTrad: 0.6276595744680851,
    i18nCode: "ps",
    langueFr: "Pachto",
  },
  {
    _id: "5ce57c969aadae8734c7aee8",
    avancementTrad: 0.5106382978723404,
    i18nCode: "fa",
    langueFr: "Persan/Dari",
  },
  {
    _id: "5ce57c969aadae8734c7aedf",
    avancementTrad: 0.7056737588652482,
    i18nCode: "en",
    langueFr: "Anglais",
  },
  {
    _id: "5db9684e5cdcfc00165352be",
    avancementTrad: 0.46099290780141844,
    i18nCode: "ti-ER",
    langueFr: "Tigrinya",
  },
  {
    _id: "5ce57c969aadae8734c7aee2",
    avancementTrad: 0.8439716312056738,
    i18nCode: "ru",
    langueFr: "Russe",
  },
  {
    _id: "5ce57c969aadae8734c7aede",
    avancementTrad: 0.9609929078014184,
    i18nCode: "ar",
    langueFr: "Arabe",
  },
];
