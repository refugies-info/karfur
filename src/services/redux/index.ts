export * from "./SelectedContent";
export * from "./Themes";
export * from "./User";
