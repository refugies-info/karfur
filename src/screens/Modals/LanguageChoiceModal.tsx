import { StyleSheet, View } from "react-native";
import Modal from "react-native-modal";
import React from "react";
import styled from "styled-components/native";
import { useDispatch, useSelector } from "react-redux";
import { saveSelectedLanguageActionCreator } from "../../services/redux/User/user.actions";
import { theme } from "../../theme";
import {
  StyledTextNormalBold,
  StyledTextSmallBold,
  StyledTextSmall,
} from "../../components/StyledText";
import { activatedLanguages } from "../../data/languagesData";
import { RowContainer } from "../../components/BasicComponents";
import { Flag } from "../../components/Language/Flag";
import { selectedI18nCodeSelector } from "../../services/redux/User/user.selectors";
import { useTranslationWithRTL } from "../../hooks/useTranslationWithRTL";
import { CustomButton } from "../../components/CustomButton";
import { AvailableLanguageI18nCode } from "../../types/interface";
interface Props {
  isModalVisible: boolean;
  toggleModal: () => void;
}

const styles = StyleSheet.create({
  view: {
    justifyContent: "flex-end",
    margin: 0,
  },
});
const MainContainer = styled.TouchableOpacity`
  padding: ${theme.margin * 2}px;
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: space-between;
  flex-wrap: wrap;
  background-color: ${(props: { isSelected: boolean }) =>
    props.isSelected ? theme.colors.black : theme.colors.lightGrey};
  border-radius: ${theme.radius * 2}px;
`;

const StyledTextBold = styled(StyledTextSmallBold)`
  text-align: left;
  margin-left: ${theme.margin}px;
  color: ${(props: { isSelected: boolean }) =>
    props.isSelected ? theme.colors.white : theme.colors.black};
`;

const StyledText = styled(StyledTextSmall)`
  text-align: left;
  color: ${(props: { isSelected: boolean }) =>
    props.isSelected ? theme.colors.white : theme.colors.black};
`;

const ModalView = styled.View`
  background-color: ${theme.colors.lightGrey};
  display: flex;
  padding-vertical: ${theme.margin}px;
  padding-horizontal: ${theme.margin * 3}px;
  border-top-right-radius: ${theme.radius * 2}px;
  border-top-left-radius: ${theme.radius * 2}px;
`;

const TitleText = styled(StyledTextNormalBold)`
  margin-top: ${theme.margin * 3}px;
  align-self: center;
`;

const LanguagesContainer = styled.ScrollView`
  margin-top: ${theme.margin * 2}px;
  margin-bottom: ${theme.margin}px;
`;

const Separator = styled.View`
  height: 1px;
  background-color: ${theme.colors.grey};
  margin-left: ${theme.margin * 2}px;
  margin-right: ${theme.margin * 2}px;
`;

const FlagBackground = styled.View`
  margin: 4px;
  background-color: ${theme.colors.white};
  width: 22px;
  height: 17px;
  display: flex;
  justify-content: center;
  align-items: center;
  border-radius: 4px;
`;
export const LanguageChoiceModal = (props: Props) => {
  const { t, i18n } = useTranslationWithRTL();
  const selectedLanguageI18nCode = useSelector(selectedI18nCodeSelector);
  const dispatch = useDispatch();

  const changeLanguage = (ln: AvailableLanguageI18nCode) => {
    i18n.changeLanguage(ln);
    dispatch(
      saveSelectedLanguageActionCreator({
        langue: ln,
        shouldFetchContents: true,
      })
    );
    props.toggleModal();
  };
  return (
    <Modal
      isVisible={props.isModalVisible}
      style={styles.view}
      onBackdropPress={props.toggleModal}
    >
      <ModalView>
        <TitleText>{t("Langue app", "Langue de l'application")}</TitleText>
        <LanguagesContainer>
          {activatedLanguages.map((language, index) => {
            const isSelected = selectedLanguageI18nCode === language.i18nCode;
            return (
              <View key={language.langueFr}>
                <MainContainer
                  onPress={() => changeLanguage(language.i18nCode)}
                  testID={"test-language-button-" + language.langueFr}
                  isSelected={isSelected}
                >
                  <RowContainer>
                    <FlagBackground>
                      <Flag langueFr={language.langueFr} />
                    </FlagBackground>
                    <StyledTextBold isSelected={isSelected}>
                      {language.langueFr}
                    </StyledTextBold>
                    {language.langueFr !== "Fran√ßais" && (
                      <RowContainer>
                        <StyledText isSelected={isSelected}>{" - "}</StyledText>
                        <StyledText isSelected={isSelected}>
                          {language.langueLoc}
                        </StyledText>
                      </RowContainer>
                    )}
                  </RowContainer>
                </MainContainer>
                {index !== activatedLanguages.length - 1 && <Separator />}
              </View>
            );
          })}
        </LanguagesContainer>
        <CustomButton
          i18nKey={"Fermer"}
          defaultText="Fermer"
          textColor={theme.colors.black}
          onPress={props.toggleModal}
          isTextNotBold={true}
        />
      </ModalView>
    </Modal>
  );
};
