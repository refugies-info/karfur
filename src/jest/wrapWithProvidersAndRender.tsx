import React from "react";
import { render } from "react-native-testing-library";
import { Provider } from "react-redux";
import { createStore } from "redux";
import {
  initialRootStateFactory,
  rootReducer,
  RootState,
} from "../services/redux/reducers";

interface WrapWithProvidersAndRenderParams {
  Component: React.FunctionComponent<unknown>;
  compProps?: Record<string, unknown>;
  reduxState?: Partial<RootState>;
}

/**
 * Provide all the ugly-to-set-up providers for your component to be ready to test
 * @param Component
 * @param compProps properties passed on to [getPropsWithNavigation()] to gen navigation props
 * @param reduxState defaults to initialRootState
 */
export function wrapWithProvidersAndRender({
  Component,
  compProps = {},
  reduxState = initialRootStateFactory(),
}: WrapWithProvidersAndRenderParams) {
  const store = createStore(rootReducer, reduxState);

  const componentWithReduxAndI18n = (
    <Provider store={store}>
      <Component {...compProps} />
    </Provider>
  );

  return render(componentWithReduxAndI18n);
}
