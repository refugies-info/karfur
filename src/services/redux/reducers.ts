import { Action, combineReducers } from "redux";
import {
  loadingStatusReducer,
  LoadingStatusState,
  initialLoadingState,
} from "./LoadingStatus/loadingStatus.reducer";
import {
  languagesReducer,
  LanguageState,
  initialLanguageState,
} from "./Languages/languages.reducer";
import { userReducer, UserState, initialUserState } from "./User/user.reducer";
import {
  contentsReducer,
  ContentsState,
  initialContentsState,
} from "./Contents/contents.reducer";
import {
  SelectedContentState,
  initialSelectedContentState,
  selectedContentReducer,
} from "./SelectedContent/selectedContent.reducer";
import {
  NeedState,
  initialNeedState,
  needsReducer,
} from "./Needs/needs.reducer";
import {
  groupedContentsReducer,
  GroupedContentsState,
  initialGroupedContentsState,
} from "./ContentsGroupedByNeeds/contentsGroupedByNeeds.reducer";
import {
  VoiceOverState,
  initialVoiceOverState,
  voiceOverReducer,
} from "./VoiceOver/voiceOver.reducer";
const appReducer = combineReducers({
  loadingStatus: loadingStatusReducer,
  languages: languagesReducer,
  user: userReducer,
  contents: contentsReducer,
  selectedContent: selectedContentReducer,
  needs: needsReducer,
  groupedContents: groupedContentsReducer,
  voiceOver: voiceOverReducer,
});

export interface RootState {
  loadingStatus: LoadingStatusState;
  languages: LanguageState;
  user: UserState;
  contents: ContentsState;
  selectedContent: SelectedContentState;
  needs: NeedState;
  groupedContents: GroupedContentsState;
  voiceOver: VoiceOverState;
}

type RootReducer = ReturnType<typeof appReducer>;

export const initialRootStateFactory = (): RootState => ({
  loadingStatus: initialLoadingState,
  languages: initialLanguageState,
  user: initialUserState,
  contents: initialContentsState,
  selectedContent: initialSelectedContentState,
  needs: initialNeedState,
  groupedContents: initialGroupedContentsState,
  voiceOver: initialVoiceOverState,
});

export const rootReducer = (
  state: RootState | undefined,
  action: Action<any>
): RootReducer =>
  // @ts-ignore
  appReducer(state, action);
