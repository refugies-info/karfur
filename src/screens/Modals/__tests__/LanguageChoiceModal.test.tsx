import { wrapWithProvidersAndRender } from "../../../jest/wrapWithProvidersAndRender";
import { initialRootStateFactory } from "../../../services/redux/reducers";
import { fireEvent, act } from "react-native-testing-library";
import { saveSelectedLanguageActionCreator } from "../../../services/redux/User/user.actions";
import { mockedLanguageData } from "../../../jest/__fixtures__/languages";
import { LanguageChoiceModal } from "../LanguageChoiceModal";
import { useTranslationWithRTL } from "../../../hooks/useTranslationWithRTL";

jest.mock("../../../services/redux/User/user.actions", () => {
  const actions = jest.requireActual(
    "../../../services/redux/User/user.actions"
  );
  return {
    saveSelectedLanguageActionCreator: jest.fn(
      actions.saveSelectedLanguageActionCreator
    ),
  };
});

jest.mock("../../../hooks/useTranslationWithRTL", () => ({
  useTranslationWithRTL: jest.fn().mockReturnValue({
    i18n: { changeLanguage: jest.fn() },
    t: jest.fn(),
  }),
}));

describe("LanguageChoiceModal", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("should render correctly", () => {
    const changeLanguage = jest.fn();
    (useTranslationWithRTL as jest.Mock).mockReturnValueOnce({
      i18n: { changeLanguage },
      t: jest.fn().mockImplementationOnce((arg1, _) => arg1),
    });
    const toggleModal = jest.fn();
    const component = wrapWithProvidersAndRender({
      Component: LanguageChoiceModal,
      reduxState: {
        ...initialRootStateFactory(),
        languages: { availableLanguages: mockedLanguageData },
      },
      compProps: { toggleModal },
    });
    expect(component).toMatchSnapshot();
    const Button = component.getByTestId("test-language-button-Anglais");
    act(() => {
      fireEvent.press(Button);
    });
    expect(changeLanguage).toHaveBeenCalledWith("en");
    expect(saveSelectedLanguageActionCreator).toHaveBeenCalledWith({
      langue: "en",
      shouldFetchContents: true,
    });
    expect(toggleModal).toHaveBeenCalledWith();
  });
});
