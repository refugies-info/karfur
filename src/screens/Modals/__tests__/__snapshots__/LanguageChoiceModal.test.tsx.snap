// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LanguageChoiceModal should render correctly 1`] = `null`;
